### ISSUE TITLE
Overloading __rmul__() on a class inheriting from int doesn't produce the correct type in Pyright

### ISSUE THREAD

**Issue by henrikhestnes** [microsoft/pyright#7046]
**Describe the bug**
Python and Pyright evaluate types differently when multiplying an `int` and a class that inherits from `int` with an overrided `__rmul__()` function.

**Code or Screenshots**
The following code in foo.py
```python
from __future__ import annotations

import typing as ty

# pyright: basic

class Foo(int):
    def __mul__(self, other: ty.SupportsFloat) -> Foo:
        if not isinstance(other, ty.SupportsFloat):
            return NotImplemented
        return Foo(int(self) * float(other))

    def __rmul__(self, other: ty.SupportsFloat) -> Foo:
        if not isinstance(other, ty.SupportsFloat):
            return NotImplemented
        return Foo(float(other) * int(self))


a: Foo = 3 * Foo(3)
b: Foo = Foo(3) * 3

print(type(a))
print(type(b))
```
produces the following output
```
<class '__main__.Foo'>
<class '__main__.Foo'>
```

However, running `pyright foo.py` in the command-line produces the following output
```
foo.py:19:10 - error: Expression of type "int" cannot be assigned to declared type "Foo"
    "int" is incompatible with "Foo" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

I can't see any other reason for this than that Pyright and Python evaluate types differently when a class is inheriting from `int` and overriding the `__rmul__()` function, which means this must be a bug in Pyright.

![image](https://github.com/microsoft/pyright/assets/59889655/afa36b4a-689d-4b59-b2db-6170017d204f)

**VS Code extension or command-line**
No extension. Running Pyright v1.1.347 (latest).

**Comment by erictraut** [comment]
Pyright's behavior is as designed here. 

The problem is that static type checkers do not have enough information to faithfully replicate the behavior of the algorithm that Python's runtime uses to determine whether to use `__mul__` or `__rmul__` in all cases. That algorithm normally uses the LHS operand's `__mul__` first and then falls back to the RHS operand's `__rmul__` if the first call returns `NotImplemented`. However, there is an exception to this rule:

> If `B <: A`, Python tries `B.__radd__` before `A.__add__`. 

The problem here is that a type checker doesn't know if `B <: A` because it doesn't know the exact object types at runtime. That's because type annotations in the Python type system represent sets of types, not exact types. For example, the type annotation `SupportsFloat` describes a protocol class that is compatible with a set of types that includes `int` and `float`.

Because of this ambiguity, pyright is unable to predict in all cases whether `__mul__` or `__rmul__` will be used at runtime, and it falls back on the normal rule that the LHS `__mul__` is used as long as the input argument types are compatible.

**Comment by gablank** [comment]
> The problem here is that a type checker doesn't know if B <: A because it doesn't know the exact object types at runtime.
Isn't this true for all type hinting? Also, shouldn't there definitely be cases where you know that `B <: A` and therefore can check the reverse method on B before the normal method on A? Shouldn't this be trivial in @henrikhestnes' example? You know that the LHS is an `int`, that the RHS is a `Foo`, and that `Foo <: int`, right?
